### ISSUE TITLE
Error `double free or corruption` when trying to decrypt drive

### ISSUE THREAD

**Issue by NTICompass** [Aorimn/dislocker#69]
I have Windows 10 and Arch Linux dual-booting on my machine.  I had Windows 10 encrypt its partition with Bitlocker (XTS-AES).

In Arch Linux, I installed dislocker 0.6.1 from https://aur.archlinux.org/packages/dislocker/

I tried to use `dislocker-fuse` to decrypt my drive so I could mount it, but I get the following output:

```
nticompass@archlinuxpc ~ % sudo dislocker-fuse -f XXX-YYY-ZZZ.BEK -vvvv -V /dev/sda2 -- /mnt/tmp
Fri Apr  8 11:27:24 2016 [DEBUG] Verbosity level to DEBUG (4) into 'stdout'
Fri Apr  8 11:27:24 2016 [INFO] dislocker by Romain Coltel, v0.6.1 (compiled for Linux/x86_64)
Fri Apr  8 11:27:24 2016 [INFO] Compiled version: :
Fri Apr  8 11:27:24 2016 [DEBUG] --- Config...
Fri Apr  8 11:27:24 2016 [DEBUG]    Verbosity: 4
Fri Apr  8 11:27:24 2016 [DEBUG]    Trying to decrypt '/dev/sda2'
Fri Apr  8 11:27:24 2016 [DEBUG]    	using the bek file at 'XXX-YYY-ZZZ.BEK'
Fri Apr  8 11:27:24 2016 [DEBUG]    Using the first valid metadata block
Fri Apr  8 11:27:24 2016 [DEBUG] ... End config ---
Fri Apr  8 11:27:24 2016 [DEBUG] Trying to open '/dev/sda2'...
Fri Apr  8 11:27:24 2016 [DEBUG] Trying to open '/dev/sda2'...
Fri Apr  8 11:27:24 2016 [DEBUG] Opened (fd #3).
Fri Apr  8 11:27:24 2016 [DEBUG] Opened (fd #3).
Fri Apr  8 11:27:24 2016 [DEBUG] New memory allocation at 0x562d45e8aee0 (0x18 bytes allocated)
Fri Apr  8 11:27:24 2016 [DEBUG] New memory allocation at 0x562d45e8af00 (0x90 bytes allocated)
Fri Apr  8 11:27:24 2016 [DEBUG] New memory allocation at 0x562d45e8afa0 (0x200 bytes allocated)
Fri Apr  8 11:27:24 2016 [DEBUG] Positionnong #3 at offset 0 from 0
Fri Apr  8 11:27:24 2016 [DEBUG] Reading volume header...
Fri Apr  8 11:27:24 2016 [DEBUG] Reading 0x200 bytes from #3 into 0x562d45e8afa0
Fri Apr  8 11:27:25 2016 [DEBUG] Volume header read
Fri Apr  8 11:27:25 2016 [DEBUG] =====[ Volume header informations ]=====
Fri Apr  8 11:27:25 2016 [DEBUG]   Signature: '-FVE-FS-'
Fri Apr  8 11:27:25 2016 [DEBUG]   Sector size: 0x0200 (512) bytes
Fri Apr  8 11:27:25 2016 [DEBUG]   Sector per cluster: 0x08 (8) bytes
Fri Apr  8 11:27:25 2016 [DEBUG]   Reserved clusters: 0x0000 (0) bytes
Fri Apr  8 11:27:25 2016 [DEBUG]   Fat count: 0x00 (0) bytes
Fri Apr  8 11:27:25 2016 [DEBUG]   Root entries: 0x0000 (0) bytes
Fri Apr  8 11:27:25 2016 [DEBUG]   Number of sectors (16 bits): 0x0000 (0) bytes
Fri Apr  8 11:27:25 2016 [DEBUG]   Media descriptor: 0xf8 (248) bytes
Fri Apr  8 11:27:25 2016 [DEBUG]   Sectors per fat: 0x0000 (0) bytes
Fri Apr  8 11:27:25 2016 [DEBUG]   Hidden sectors: 0x00032800 (206848) bytes
Fri Apr  8 11:27:25 2016 [DEBUG]   Number of sectors (32 bits): 0x00000000 (0) bytes
Fri Apr  8 11:27:25 2016 [DEBUG]   Number of sectors (64 bits): 0x0000000000000000 (0) bytes
Fri Apr  8 11:27:25 2016 [DEBUG]   MFT start cluster: 0x0000000000060001 (393217) bytes
Fri Apr  8 11:27:25 2016 [DEBUG]   Metadata Lcn: 0x0000000000000000 (0) bytes
Fri Apr  8 11:27:25 2016 [DEBUG]   Volume GUID: 'XYZ-XYZ-XYZ'
Fri Apr  8 11:27:25 2016 [DEBUG]   First metadata header offset:  0x00000000062a0000
Fri Apr  8 11:27:25 2016 [DEBUG]   Second metadata header offset: 0x000000004eae7000
Fri Apr  8 11:27:25 2016 [DEBUG]   Third metadata header offset:  0x00000000857b4000
Fri Apr  8 11:27:25 2016 [DEBUG]   Boot Partition Identifier: '0xaa55'
Fri Apr  8 11:27:25 2016 [DEBUG] ========================================
Fri Apr  8 11:27:25 2016 [INFO] Volume has EOW_INFORMATION_OFFSET_GUID.
Fri Apr  8 11:27:25 2016 [DEBUG] Positionnong #3 at offset 88739840 from 0
Fri Apr  8 11:27:25 2016 [DEBUG] Reading EOW Information header at 0x54a1000...
Fri Apr  8 11:27:25 2016 [DEBUG] Reading 0x38 bytes from #3 into 0x7ffceaacb390
Fri Apr  8 11:27:25 2016 [DEBUG] New memory allocation at 0x562d45e8b1b0 (0x3b8 bytes allocated)
Fri Apr  8 11:27:25 2016 [DEBUG] Reading EOW information's payload...
Fri Apr  8 11:27:25 2016 [DEBUG] Reading 0x348 bytes from #3 into 0x562d45e8b1e8
Fri Apr  8 11:27:25 2016 [DEBUG] End get_eow_information.
Fri Apr  8 11:27:25 2016 [DEBUG] =======================[ BitLocker EOW informations ]========================
Fri Apr  8 11:27:25 2016 [DEBUG]   Signature: 'FVE-EOW'
Fri Apr  8 11:27:25 2016 [DEBUG]   Structure size: 0x0038 (56)
Fri Apr  8 11:27:25 2016 [DEBUG]   On-disk size: 0x03b8 (952)
Fri Apr  8 11:27:25 2016 [DEBUG]   Sector size (1): 0x0200 (512)
Fri Apr  8 11:27:25 2016 [DEBUG]   Sector size (2): 0x0200 (512)
Fri Apr  8 11:27:25 2016 [DEBUG]   Unknown (0x14): 0x00100000 (1048576)
Fri Apr  8 11:27:25 2016 [DEBUG]   Convlog size: 0x00010c00 (68608)
Fri Apr  8 11:27:25 2016 [DEBUG]   Unknown (0x1c): 0x00008400 (33792)
Fri Apr  8 11:27:25 2016 [DEBUG]   Number of regions: 112
Fri Apr  8 11:27:25 2016 [DEBUG]   Crc32: 46971904
Fri Apr  8 11:27:25 2016 [DEBUG]   On-disk offsets: 0x562d45e8b1d8
Fri Apr  8 11:27:25 2016 [DEBUG] ============================================================================
=
Fri Apr  8 11:27:25 2016 [DEBUG] Freeing pointer at address 0x562d45e8b1b0
Fri Apr  8 11:27:25 2016 [DEBUG] Entering get_eow_check_valid
Fri Apr  8 11:27:25 2016 [DEBUG] Positionnong #3 at offset 88739840 from 0
Fri Apr  8 11:27:25 2016 [DEBUG] Reading EOW Information header at 0x54a1000...
Fri Apr  8 11:27:25 2016 [DEBUG] Reading 0x38 bytes from #3 into 0x7ffceaacb390
Fri Apr  8 11:27:25 2016 [DEBUG] New memory allocation at 0x562d45e8b1b0 (0x3b8 bytes allocated)
Fri Apr  8 11:27:25 2016 [DEBUG] Reading EOW information's payload...
Fri Apr  8 11:27:25 2016 [DEBUG] Reading 0x348 bytes from #3 into 0x562d45e8b1e8
Fri Apr  8 11:27:25 2016 [DEBUG] End get_eow_information.
Fri Apr  8 11:27:25 2016 [DEBUG] Looking if 0xf38377b3 == 0x46971904 for EOW information validation
Fri Apr  8 11:27:25 2016 [DEBUG] Freeing pointer at address 0x562d45e8b1b0
Fri Apr  8 11:27:25 2016 [DEBUG] Positionnong #3 at offset 1702043648 from 0
Fri Apr  8 11:27:25 2016 [DEBUG] Reading EOW Information header at 0x65732000...
Fri Apr  8 11:27:25 2016 [DEBUG] Reading 0x38 bytes from #3 into 0x7ffceaacb390
Fri Apr  8 11:27:25 2016 [DEBUG] New memory allocation at 0x562d45e8b1b0 (0x3b8 bytes allocated)
Fri Apr  8 11:27:25 2016 [DEBUG] Reading EOW information's payload...
Fri Apr  8 11:27:25 2016 [DEBUG] Reading 0x348 bytes from #3 into 0x562d45e8b1e8
Fri Apr  8 11:27:25 2016 [DEBUG] End get_eow_information.
Fri Apr  8 11:27:25 2016 [DEBUG] Looking if 0xf38377b3 == 0x46971904 for EOW information validation
Fri Apr  8 11:27:25 2016 [DEBUG] Freeing pointer at address 0x562d45e8b1b0
Fri Apr  8 11:27:25 2016 [INFO] EOW information at offset 54a1000 passed the tests
Fri Apr  8 11:27:25 2016 [DEBUG] Freeing pointer at address 0x562d45e8b1b0
*** Error in `dislocker-fuse': double free or corruption (top): 0x0000562d45e8b1b0 ***
======= Backtrace: =========
/usr/lib/libc.so.6(+0x6f364)[0x7f90498cd364]
/usr/lib/libc.so.6(+0x74d96)[0x7f90498d2d96]
/usr/lib/libc.so.6(+0x7557e)[0x7f90498d357e]
/usr/lib/libdislocker.so.0.6(dis_free+0x25)[0x7f9049c06b13]
/usr/lib/libdislocker.so.0.6(dis_metadata_initialize+0xa40)[0x7f9049c094dd]
/usr/lib/libdislocker.so.0.6(dis_initialize+0x1ed)[0x7f9049c05323]
dislocker-fuse(main+0x5b)[0x562d44b87161]
/usr/lib/libc.so.6(__libc_start_main+0xf0)[0x7f904987e710]
dislocker-fuse(_start+0x29)[0x562d44b86d19]
======= Memory map: ========
562d44b86000-562d44b88000 r-xp 00000000 fe:02 2372389                    /usr/bin/dislocker-fuse
562d44d87000-562d44d88000 r--p 00001000 fe:02 2372389                    /usr/bin/dislocker-fuse
562d44d88000-562d44d89000 rw-p 00002000 fe:02 2372389                    /usr/bin/dislocker-fuse
562d45e89000-562d45eaa000 rw-p 00000000 00:00 0                          [heap]
7f9044000000-7f9044021000 rw-p 00000000 00:00 0
7f9044021000-7f9048000000 ---p 00000000 00:00 0
7f9048361000-7f9048377000 r-xp 00000000 fe:02 2228586                    /usr/lib/libgcc_s.so.1
7f9048377000-7f9048576000 ---p 00016000 fe:02 2228586                    /usr/lib/libgcc_s.so.1
7f9048576000-7f9048577000 rw-p 00015000 fe:02 2228586                    /usr/lib/libgcc_s.so.1
7f90485a9000-7f90486ac000 r-xp 00000000 fe:02 2228319                    /usr/lib/libm-2.23.so
7f90486ac000-7f90488ac000 ---p 00103000 fe:02 2228319                    /usr/lib/libm-2.23.so
7f90488ac000-7f90488ad000 r--p 00103000 fe:02 2228319                    /usr/lib/libm-2.23.so
7f90488ad000-7f90488ae000 rw-p 00104000 fe:02 2228319                    /usr/lib/libm-2.23.so
7f90488ae000-7f90488b6000 r-xp 00000000 fe:02 2228310                    /usr/lib/libcrypt-2.23.so
7f90488b6000-7f9048ab5000 ---p 00008000 fe:02 2228310                    /usr/lib/libcrypt-2.23.so
7f9048ab5000-7f9048ab6000 r--p 00007000 fe:02 2228310                    /usr/lib/libcrypt-2.23.so
7f9048ab6000-7f9048ab7000 rw-p 00008000 fe:02 2228310                    /usr/lib/libcrypt-2.23.so
7f9048ab7000-7f9048ae5000 rw-p 00000000 00:00 0
7f9048ae5000-7f9048b5c000 r-xp 00000000 fe:02 2228692                    /usr/lib/libgmp.so.10.3.0
7f9048b5c000-7f9048d5b000 ---p 00077000 fe:02 2228692                    /usr/lib/libgmp.so.10.3.0
7f9048d5b000-7f9048d5c000 r--p 00076000 fe:02 2228692                    /usr/lib/libgmp.so.10.3.0
7f9048d5c000-7f9048d5d000 rw-p 00077000 fe:02 2228692                    /usr/lib/libgmp.so.10.3.0
7f9048d5d000-7f9048fd0000 r-xp 00000000 fe:02 2240528                    /usr/lib/libruby.so.2.3.0
7f9048fd0000-7f90491cf000 ---p 00273000 fe:02 2240528                    /usr/lib/libruby.so.2.3.0
7f90491cf000-7f90491d7000 rw-p 00272000 fe:02 2240528                    /usr/lib/libruby.so.2.3.0
7f90491d7000-7f90491e8000 rw-p 00000000 00:00 0
7f90491e8000-7f9049237000 r-xp 00000000 fe:02 2240531                    /usr/lib/libmbedcrypto.so.0
7f9049237000-7f9049436000 ---p 0004f000 fe:02 2240531                    /usr/lib/libmbedcrypto.so.0
7f9049436000-7f9049439000 r--p 0004e000 fe:02 2240531                    /usr/lib/libmbedcrypto.so.0
7f9049439000-7f904943a000 rw-p 00051000 fe:02 2240531                    /usr/lib/libmbedcrypto.so.0
7f904943a000-7f904943d000 rw-p 00000000 00:00 0
7f904943d000-7f9049455000 r-xp 00000000 fe:02 2228242                    /usr/lib/libpthread-2.23.so
7f9049455000-7f9049654000 ---p 00018000 fe:02 2228242                    /usr/lib/libpthread-2.23.so
7f9049654000-7f9049655000 r--p 00017000 fe:02 2228242                    /usr/lib/libpthread-2.23.so
7f9049655000-7f9049656000 rw-p 00018000 fe:02 2228242                    /usr/lib/libpthread-2.23.so
7f9049656000-7f904965a000 rw-p 00000000 00:00 0
7f904965a000-7f904965c000 r-xp 00000000 fe:02 2228318                    /usr/lib/libdl-2.23.so
7f904965c000-7f904985c000 ---p 00002000 fe:02 2228318                    /usr/lib/libdl-2.23.so
7f904985c000-7f904985d000 r--p 00002000 fe:02 2228318                    /usr/lib/libdl-2.23.so
7f904985d000-7f904985e000 rw-p 00003000 fe:02 2228318                    /usr/lib/libdl-2.23.so
7f904985e000-7f90499f6000 r-xp 00000000 fe:02 2228261                    /usr/lib/libc-2.23.so
7f90499f6000-7f9049bf5000 ---p 00198000 fe:02 2228261                    /usr/lib/libc-2.23.so
7f9049bf5000-7f9049bf9000 r--p 00197000 fe:02 2228261                    /usr/lib/libc-2.23.so
7f9049bf9000-7f9049bfb000 rw-p 0019b000 fe:02 2228261                    /usr/lib/libc-2.23.so
7f9049bfb000-7f9049bff000 rw-p 00000000 00:00 0
7f9049bff000-7f9049c18000 r-xp 00000000 fe:02 2240540                    /usr/lib/libdislocker.so.0.6.1
7f9049c18000-7f9049e17000 ---p 00019000 fe:02 2240540                    /usr/lib/libdislocker.so.0.6.1
7f9049e17000-7f9049e18000 r--p 00018000 fe:02 2240540                    /usr/lib/libdislocker.so.0.6.1
7f9049e18000-7f9049e19000 rw-p 00019000 fe:02 2240540                    /usr/lib/libdislocker.so.0.6.1
7f9049e19000-7f9049e44000 r-xp 00000000 fe:02 2233715                    /usr/lib/libfuse.so.2.9.4
7f9049e44000-7f904a044000 ---p 0002b000 fe:02 2233715                    /usr/lib/libfuse.so.2.9.4
7f904a044000-7f904a056000 r--p 0002b000 fe:02 2233715                    /usr/lib/libfuse.so.2.9.4
7f904a056000-7f904a057000 rw-p 0003d000 fe:02 2233715                    /usr/lib/libfuse.so.2.9.4
7f904a057000-7f904a07a000 r-xp 00000000 fe:02 2228260                    /usr/lib/ld-2.23.so
7f904a0a9000-7f904a241000 r--p 00000000 fe:02 2233713                    /usr/lib/locale/locale-archive
7f904a241000-7f904a248000 rw-p 00000000 00:00 0
7f904a279000-7f904a27a000 rw-p 00000000 00:00 0
7f904a27a000-7f904a27b000 r--p 00023000 fe:02 2228260                    /usr/lib/ld-2.23.so
7f904a27b000-7f904a27c000 rw-p 00024000 fe:02 2228260                    /usr/lib/ld-2.23.so
7f904a27c000-7f904a27d000 rw-p 00000000 00:00 0
7ffceaaab000-7ffceaacc000 rw-p 00000000 00:00 0                          [stack]
7ffceabb2000-7ffceabb4000 r--p 00000000 00:00 0                          [vvar]
7ffceabb4000-7ffceabb6000 r-xp 00000000 00:00 0                          [vdso]
ffffffffff600000-ffffffffff601000 r-xp 00000000 00:00 0                  [vsyscall]
zsh: abort      sudo dislocker-fuse -f XXX-YYY-ZZZ.BEK -vvvv -V  --
```

**Comment by Aorimn** [comment]
Hi,
Support for EOW volumes is not great (as you can see) because I didn't have any test data. I'm very interested in increasing the support for these volumes and fix the above bug.

Can you explain how did you create your encrypted volume? What's the filesystem on the volume? Which volume type is it (C: drive, data drive, external drive)?

**Comment by NTICompass** [comment]
This is my `C:` drive, it's NTFS.  I just right-clicked on it in My Computer and enabled bitlocker.  It asked whether I wanted to encrypt the whole drive or just used space and then asked what mode.

I selected the whole drive and set it to XTS-AES mode.

**Comment by Aorimn** [comment]
If it's your `C:` drive, do you know if the TPM is used?
If you didn't touch to gpedits.msc settings, then it should be the case, AFAIK.

**Comment by NTICompass** [comment]
I did edit the group policy, actually.  My PC doesn't have a TPM chip, so I edited the policy to let me use BitLocker on the system drive without it.

**Comment by Aorimn** [comment]
So only this policy?

**Comment by NTICompass** [comment]
The only policy I edited was "Require additional authentication at startup" (found at http://www.howtogeek.com/howto/6229/how-to-use-bitlocker-on-drives-without-tpm/).

I noticed something when I booted up Windows earlier.  It was *still* encrypting the files on the drive!  I'm going to leave it overnight to let it finish.  I usually use Linux, and Windows didn't actually tell me that it wasn't 100% finished.
